src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -386,34 +410,61 @@ public override int GetHashCode()
         {
-            ElementTypeWithAnnotations.AddNullableTransforms(transforms);
-        }
-
-        internal override bool ApplyNullableTransforms(byte defaultTransformFlag, ImmutableArray<byte> transforms, ref int position, out TypeSymbol result)
-        {
-            TypeWithAnnotations oldElementType = ElementTypeWithAnnotations;
-            TypeWithAnnotations newElementType;
-
-            if (!oldElementType.ApplyNullableTransforms(defaultTransformFlag, transforms, ref position, out newElementType))
+            var elementType = ElementTypeWithAnnotations;
+            while (elementType.Type is ArrayTypeSymbol arrayType)
             {
-                result = this;
-                return false;
+                elementType.AddNullableTransformShallow(transforms);
+                elementType = arrayType.ElementTypeWithAnnotations;
             }
 
-            result = WithElementType(newElementType);
-            return true;
+            elementType.AddNullableTransforms(transforms);
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
-        {
-            return WithElementType(transform(ElementTypeWithAnnotations));
-        }
+        internal override TypeSymbol ApplyNullableTransforms(NullableTransformStream stream) =>
+            TransformNullable(ApplyTransformNullableTransform.Instance, stream);
+
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes() =>
+            TransformNullable(SetObliviousNullableTransform.Instance, null);
 
         internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind variance)
         {
             Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
-            TypeWithAnnotations elementType = ElementTypeWithAnnotations.MergeNullability(((ArrayTypeSymbol)other).ElementTypeWithAnnotations, variance);
-            return WithElementType(elementType);
+            return TransformNullable(MergeNullableTransform.Instance, ((ArrayTypeSymbol)other, variance));
+        }
+
+        private ArrayTypeSymbol TransformNullable<TState, TInitialState>(NullableTransform<TState, TInitialState> transform, TInitialState initialState)
+        {
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach to building up the annotations here.
+            var builder = ArrayBuilder<TState>.GetInstance();
+
+            // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
+            // as oblivious
+            var array = this;
+            var state = transform.GetInitialState(initialState, this);
+            builder.Push(state);
+            do
+            {
```

Comment:
I believe this transformation (to an iterative solution) is correct, but it is hard to read, understand, and perhaps maintain.  I prefer your previous iteration as being a simpler approach.

Reply:
I agree, it also doesn't look like this involves significantly less code by comparison to duplicating stack approach in each method that needs it, the would be duplicated portions of code look small by comparison the the specialized logic, which is now disjoined from the algorithm and moved into individual classes. Makes it hard to reason about what is going on. 

---
In reply to: [274627547](https://github.com/dotnet/roslyn/pull/34891#discussion_r274627547) [](ancestors = 274627547)

Reply:
I've gone back and forth on this. The original solution looked simpler because it was the easiest case to consider. The other ones, specifically `ApplyNullableTransform` are less readable. The algorithm gets lost in the transform. 

Let me essentially undo this work and inline it in a single commit. That will give reviewers both solution and we can decide which we prefer to use going forward. 

---
In reply to: [274711575](https://github.com/dotnet/roslyn/pull/34891#discussion_r274711575) [](ancestors = 274711575,274627547)

Reply:
Done. If you diff iterations 15 and 16 you can contrast the two approaches here.

---
In reply to: [274716369](https://github.com/dotnet/roslyn/pull/34891#discussion_r274716369) [](ancestors = 274716369,274711575,274627547)

Reply:
Should we have a `TypeVisitor` base class that can be used for many scenarios where we visit or rewrite types within a compound type?

The base class would hide any complexity around deeply-nested types.

If the `TypeVisitor` methods take extra args, hopefully visitors could be implemented with singleton instances rather than requiring allocations for the visitors.
Ideally, the `TypeVisitor` class could also replace the `VisitType<T>` extension method we have today.

For instance:

```C#
abstract class TypeVisitor<TArg, TResult>
{
    internal abstract (TResult, TypeWithAnnotations) Visit(TypeWithAnnotations type, TArg arg);

    internal abstract (TResult, TypeSymbol) Visit(TypeSymbol type, TArg arg);

    protected (TResult, TypeSymbol) DefaultVisit(TypeSymbol type, TArg arg)
    {
        // ...
    }
}

sealed class ApplyNullableTransformsVisitor : TypeVisitor<object, NullableTransformStream>
{
}

sealed class SetObliviousNullabilityVisitor : TypeVisitor<object, object>
{
}

sealed class MergeNullabilityVisitor : TypeVisitor<object, VarianceKind>
{
}

// Example that uses VisitType<T> extension method today
sealed class IsContainingSymbolOfAllTypeParametersVisitor : TypeVisitor<Symbol, bool>
{
}
```